README.md
# Laravel Inplace

[![Latest Version on Packagist](https://img.shields.io/packagist/v/devsrv/laravel-inplace.svg?style=flat-square)](https://packagist.org/packages/devsrv/laravel-inplace)
[![Total Downloads](https://img.shields.io/packagist/dt/devsrv/laravel-inplace.svg?style=flat-square)](https://packagist.org/packages/devsrv/laravel-inplace)

Because forms are boaring.

This package provide set of blade components to make editing content exciting & blazing fast.

### Installation

```shell
composer require devsrv/laravel-inplace
```

### Setup (_required_)

include `@include('inplace::styles')` and `@include('inplace::scripts')` on every page where you'll be using the _inplace_ component

> its best to add those directives in your main layout blade file and extend that blade layout

```php
...
    @include('inplace::styles')
</head>
<body>
    ...

    @include('inplace::scripts')
</body>
</html>
```

### Publishing Frontend Assets (_required_)
```shell
php artisan vendor:publish --provider="devsrv\inplace\InplaceServiceProvider" --tag=public
```

##### 💡 NOTE :
> when updating the package make sure to use the `--force` flag to keep the assets up-to-date i.e. 
> 
> `php artisan vendor:publish --provider="devsrv\inplace\InplaceServiceProvider" --tag=public --force`

### Publish config (_optional_)

```shell
php artisan vendor:publish --provider="devsrv\inplace\InplaceServiceProvider" --tag=config
```
#### ✔️ icons :
svg content for `edit` `save` and `cancel` button icon

#### ✔️ middleware : 
add as many middlewares you with in the `middleware` array e.g.: `['auth', 'admin']`



## Guide

### Inline Editable [All Examples here](https://github.com/devsrv/laravel-inplace-example/blob/master/resources/views/welcome.blade.php)

**Example 1** | simplest usage

```php
<x-inplace-text
  model="App\Models\User:name,1"		// (REQUIRED) format: App\ModelNamespace\Model:id
  :model="\App\Models\User::find(1)" 	// Alternatively you can pass model instance
  column="name"							// (OPTIONAL) name of the table column to update
  validation="required|min:10"			// (OPTIONAL) pass validation rules
>
  {{ \App\Models\User::find(1)->name }}
</x-inplace-text>
```

**Example 2** | Slotted Markup

```php
<x-inplace-text
	:model="$user"
    ...
>
   <x-slot name="before"><div class="myclass anotherclass"><h2></x-slot>	// custom markup prepend
   <x-slot name="after"></h2></div></x-slot>								// custom markup append

    {{ $user->email }}
</x-inplace-text>
```

**Example - 3** | Pass Custom Class to save content
[check here](https://github.com/devsrv/laravel-inplace-example/blob/c81c21d76c888958964b1eb1a589ea524694f8e9/resources/views/welcome.blade.php#L40)

> the custom save class must be a invokable object (class with __invoke method) which receives `$model, $column, $value` and it should return an array with key `success` (bool) and `message` (string)


_Example Custom Save class :_ refer to this [CustomSave](https://github.com/devsrv/laravel-inplace-example/blob/master/app/Http/Inplace/Requests/CustomSave.php) full example

**Example - 4** | Render content as custom component

```php
<x-inplace-text
  model="..."
  render-as="CustomInlineRender"		// pass your own blade component which takes care of how content gets rendered
>
  ...
</x-inplace-text>
```

> make sure to pass `{{ $attributes }}` to the html elenent that is wrapping the target content
> e.g.: `<h1 class="your-class" {{ $attributes }}></h1>`

refer to this example [component](https://github.com/devsrv/laravel-inplace-example/blob/master/resources/views/components/custom-inline-render.blade.php)

**Example - 5** | Complex Validation Rules

```php
@php
$rules = ['required', \Illuminate\Validation\Rule::in(['11', '12']), 'min:2'];
@endphp

<x-inplace-text
model="App\Models\User:1"
column="name"
:validation="$rules"                   // complex validation can be passed`
>
  {{ \App\Models\User::find(1)->name }}
</x-inplace-text>
```
refer [this example](https://github.com/devsrv/laravel-inplace-example/blob/2225ee785f4369f970cedaa09578d29c08b91098/resources/views/welcome.blade.php#L50)

#### 📌 Note: 
if using direct inplace blade component to pass all the configs, authorization will be enforced always. to override/customize this behaviour consider using [field maker](https://github.com/devsrv/laravel-inplace#advanced) configurator.


### 👾ADVANCED
instead passing config via attributes you can use the advanced field configurator file where you have access to fluent config setter methods, also this approach lets you reuse same config for multiple edits and more fine grained options to configure

```shell
php artisan inplace:config {all | text | relation}
```
this command will create `App\Http\Inplace\Text.php` and `App\Http\Inplace\Relation.php` or just one of them depending on your input

in the `config` method of the class add multiple configs as array 
> for inline field: using `devsrv\inplace\InlineText` class and 
> 
> for relation field: using `devsrv\inplace\RelationManager` class

then you can simply use the component as:
```php
<x-inplace-text
	id="USERNAME"							// the should match id of field config
	:model="\App\Models\User::find(1)"		// however model still needs to be passed via attribute (always required)
>
```

refer to [this file](https://github.com/devsrv/laravel-inplace-example/blob/master/app/Http/Inplace/Text.php) for example

> Field configurator supports some extra methods like: 
`authorizeUsing( closure )`, `bypassAuthorize()`, `middleware(['foo', 'bar'])` etc.

**_detailed documentation comming soon_ . . .**

#### ☄️ Rate Limiting:
there are two ways to rate limit inplace requests
1. __Generic Rate Limiter:__ define [rate limiter](https://laravel.com/docs/8.x/routing#defining-rate-limiters) and put the rate limiter middleware name in either config file to apply globally or attach in the [field config](https://github.com/devsrv/laravel-inplace-example/blob/c81c21d76c888958964b1eb1a589ea524694f8e9/app/Http/Inplace/Text.php#L18) in `->middleware()` method. [example](https://github.com/devsrv/laravel-inplace-example/blob/47ecea39a8713db0f510be320ce548b4514878df/app/Providers/RouteServiceProvider.php#L66)
> by doing this when a field's request gets blocked by 429, any field which is configured with that same rate limiter middleware will also be blocked

2. __Field Level Rate Limiter:__ define [rate limiter](https://github.com/devsrv/laravel-inplace-example/blob/47ecea39a8713db0f510be320ce548b4514878df/app/Providers/RouteServiceProvider.php#L64) using `devsrv\inplace\RateLimiter` and attach it in [field config](https://github.com/devsrv/laravel-inplace-example/blob/47ecea39a8713db0f510be320ce548b4514878df/app/Http/Inplace/Relation.php#L30)
> field level rate limiter blocks only same type fields. 

_Example 1_ - if you are saving the name for user with id 1 and the field gets rate limited then this has no effect on any other field which is configured with the same field level rate limiter middleware.

_Example 2_ - if you are updating badges relation of user with id 2 and the request gets rate limited then this has no effect on any other relation or even non relation field which is configured with the same field level rate limiter middleware

##### Defining Field Level Rate Limiter: [example](https://github.com/devsrv/laravel-inplace-example/blob/master/app/Providers/RouteServiceProvider.php#L64)
_`App\Providers\RouteServiceProvider.php`_
```php
use devsrv\inplace\RateLimiter as InplaceFieldRateLimiter;

protected function configureRateLimiting()
{
	InplaceFieldRateLimiter::for('author_badges')->perMinute(1);  // support perMinutes | perHour | perDay
}
```

Supported available limiter methods

| Method | Description                                                                            |
|----------------|----------------------------------------------------------------------------------------|
| perMinute      | [check api](https://laravel.com/api/8.x/Illuminate/Cache/RateLimiting/Limit.html#method_perMinute)  |
| perMinutes     | [check api](https://laravel.com/api/8.x/Illuminate/Cache/RateLimiting/Limit.html#method_perMinutes) |
| perHour        | [check api](https://laravel.com/api/8.x/Illuminate/Cache/RateLimiting/Limit.html#method_perHour)    |
| perDay         | [check api](https://laravel.com/api/8.x/Illuminate/Cache/RateLimiting/Limit.html#method_perDay)     |


### 🎁 Bonus:

#### 1. **authorize manually:** 
when passing custom class to save data you may choose to authorize the action from within your class using
   1. `Gate::authorize('update', $model);` OR
   2. `Gate::authorize('edit-settings');` OR
   3. `$this->authorize('update', $model);`

> donn't forget to use the `Illuminate\Foundation\Auth\Access\AuthorizesRequests` trait.

referer this [example](https://github.com/devsrv/laravel-inplace-example/blob/master/app/Http/Inplace/Requests/CustomSave.php) 

#### 2. if you use the popular [SPATIE PERMISSION](https://github.com/spatie/laravel-permission) package: 
you may choose to use `authorizeSpatieRoleOrPermission` method that comes with the package as a support when you use the `devsrv\inplace\Traits\SpatieAuthorize` trait.

```php
use devsrv\inplace\Traits\SpatieAuthorize;

class CustomSave
{
    use SpatieAuthorize;

    public function __invoke($model, $column, $value)
    {
    	$this->authorizeSpatieRoleOrPermission(['admin', 'some permission']);

        // save data here

        return [
            'success' => 0,
            'message' => 'not allowed'
        ];
    }
}
```

refer this [example](https://github.com/devsrv/laravel-inplace-example/blob/c81c21d76c888958964b1eb1a589ea524694f8e9/app/Http/Inplace/Requests/CustomSave.php#L32)

#### 2. 🔥 Listen events
1. `inplace-editable-progress` custom `window` browser event diaptched after ajax start & ajax finished. refer to [example](https://github.com/devsrv/laravel-inplace-example/blob/3057161a1af84a2f9a9c215157f0e28c9edcb1c4/resources/views/app.blade.php#L58) for [NProgress](https://github.com/rstacruz/nprogress) Implementation
2. `inplace-editable-finish` custom `window` browser event diaptched after content is either saved | failed by server. refer to [example](https://github.com/devsrv/laravel-inplace-example/blob/3057161a1af84a2f9a9c215157f0e28c9edcb1c4/resources/views/app.blade.php#L49) for a sample notifier system

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## 👋🏼 Say Hi! 
Leave a ⭐ if you find this package useful 👍🏼,
don't forget to let me know in [Twitter](https://twitter.com/srvrksh)  


CHANGELOG.md
# Changelog

## 2.1.0 - 2021-06-26

#### Added
    - support rate limiting requests - generic & field level rate limiter

#### Changed
    - validation rules no longer needed to be serialized when passing directly via component attribute

## 2.0.0 - 2021-06-13

#### Added
    - new methods to customize authorization behavioir for text component - `authorizeUsing`, `bypassAuthorize` 
    - relation component memoization for options list
    - relation component do not re-query options if relation already eager loaded
    - if data save error then relation options selection revert back
    - relation component replace current values markup after successful data save

#### Removed
    - global authorize config removed. if using blade component attribute to pass all configs then authorization will always be enforced
    - `authorize` `withoutMiddleware` config methods removed from text component 
    - `authorize` attribute no longer supported for text component

#### Changed
    - `x-inplace-inline` to `x-inplace-text`
    - `saveusing` renammed to `save-using`
    - custom save-as class support invokable object

## 1.3.1 - 2021-04-25

- edit control button appear on content hover
- config support for buttons custom svg icons
- inline editable text doubleclick support

## 1.3.0 - 2021-04-24

- instead of serviceprovider for advanced configuration now using seperate configurator files [ App\Http\Inplace\Inline.php App\Http\Inplace\Relation.php ]
- publish advanced config using: `php artisan inplace:config`
- `inplace:config` command support arguments - `all` `inline` `relation`

## 1.2.0 - 2021-04-18

inline edit
- vendor publish tag support [`public`, `config`] `php artisan vendor:publish --provider="devsrv\inplace\InplaceServiceProvider" --tag=public`
- inline edit component name changed from `x-inplace-component` to `x-inplace-inline`
- changed `authorize` attribute behaviour [true - user authorized, false - user not authorized]
- model attribute input format changed [ namespace\Model\Class:primaryKeyValue - ex. `App\Models\User:1` ]
- model attribute support model instance too
- added column attribute [the column name to update]
- advanced field configuration support via fluent setter methods
    - publish using: `php artisan inplace:config`
    - register `App\Providers\InplaceConfigServiceProvider::class` in `app.php`

## 1.1.3 - 2021-04-04

- support for middlewares via published config
- lottie animation for success & failed saving states
- error messages notification & validation errors list when 422

## 1.1.2 - 2021-04-04

- lottie json path missing fix

## 1.1.1 - 2021-04-02

- using webpack bundled script instead inline
- alpine used as npm module
- no conflict with user's alpine / alpine magic helpers as all inplace script is bundled
- lottie animation support for success notification

## 1.1.0 - 2021-03-28

- removed livewire
- custom `saveusing` & `model` attribute data encrypted using OpenSSL and the AES-256-CBC cipher
- after data save `inplace-editable-finish` event dispatched by inplace with property `success : true|false`
- data save start & end track ajax progress by listening `inplace-editable-progress` event

## 1.0.0 - 2021-03-21

- initial release

config/inplace.php
<?php

return [
    'middleware' => [],

    'icons' => [
        'edit' => '<svg class="inplace-control-icon" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"></path></svg>',
        'save' => '<svg class="inplace-control-icon" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 13l4 4L19 7"></path></svg>',
        'cancel' => '<svg class="inplace-control-icon" aria-hidden="true" focusable="false" data-prefix="far" data-icon="times-circle" class="svg-inline--fa fa-times-circle fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm101.8-262.2L295.6 256l62.2 62.2c4.7 4.7 4.7 12.3 0 17l-22.6 22.6c-4.7 4.7-12.3 4.7-17 0L256 295.6l-62.2 62.2c-4.7 4.7-12.3 4.7-17 0l-22.6-22.6c-4.7-4.7-4.7-12.3 0-17l62.2-62.2-62.2-62.2c-4.7-4.7-4.7-12.3 0-17l22.6-22.6c4.7-4.7 12.3-4.7 17 0l62.2 62.2 62.2-62.2c4.7-4.7 12.3-4.7 17 0l22.6 22.6c4.7 4.7 4.7 12.3 0 17z"></path></svg>',
    ]
];
